pkg/keystore/wd.go

```diff
@@ -0,0 +1,354 @@
+package keystore
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/go-redis/redis/v8"
+	"github.com/google/uuid"
+	"github.com/ledgerhq/bitcoin-keychain/pb/bitcoin"
+	"github.com/ledgerhq/bitcoin-keychain/pkg/chaincfg"
+)
+
+// WDKeystore implement the Keystore interface to write on "wallet daemon" redis
+// database, also known as "user pref"
+
+type WDKeystore struct {
+	db     *redis.Client
+	client bitcoin.CoinServiceClient
+}
+
+func NewWDKeystore(redisOpts *redis.Options) (*WDKeystore, error) {
+	rdb := redis.NewClient(redisOpts)
+
+	if _, err := rdb.Ping(context.Background()).Result(); err != nil {
+		return nil, fmt.Errorf("Pinging redis failed: %w", err)
+	}
+
+	return &WDKeystore{
+		db:     rdb,
+		client: bitcoin.NewBitcoinClient(),
+	}, nil
+}
```

Comment:
```suggestion
type WDKeystore struct {
	baseRedisKeystore
}

func NewWDKeystore(redisOpts *redis.Options) (*WDKeystore, error) {
	rdb := redis.NewClient(redisOpts)

	if _, err := rdb.Ping(context.Background()).Result(); err != nil {
		return nil, fmt.Errorf("Pinging redis failed: %w", err)
	}

	baseKeystore := baseRedisKeystore{
		db:     rdb,
		client: bitcoin.NewBitcoinClient(),
	}

	return &WDKeystore{baseKeystore }, nil
}
```

Then you should be able to add the following to `redis_common.go`.

```go
type baseRedisKeystore struct {
	db     *redis.Client
	client bitcoin.CoinServiceClient
}

func (s *baseRedisKeystore) Get(id uuid.UUID) (KeychainInfo, error) {
	var meta Meta

	err := get(s.db, id.String(), &meta)
	if err != nil {
		return KeychainInfo{}, ErrKeychainNotFound
	}

	return meta.Main, nil
}
```

The `client` and `db` fields on `WDKeystore` will work as usual.

This pattern gives you the benefits of inheritance but using composition. You can read about it [here](https://travix.io/type-embedding-in-go-ba40dd4264df).

Reply:
very cool, thx !